protos/idl/service.proto

```diff
@@ -98,3 +109,68 @@ message Tag {
 message Metadata {
     map<string, string> key_map = 1; // key map is a dictionary of key/val strings that represent metadata
 }
+
+message FilterExpression {
+    repeated SingleValueFilter filters = 1;
+}
+
+message SingleValueFilter {
+    oneof entity {
+        TagFilter tagFilter = 1;
+        PartitionFilter partition_filter = 2;
+        ArtifactFilter artifact_filter = 3;
+        DatasetFilter dataset_filter = 4;
+    }
+
+    // as use-cases come up we can add more operators, ex: gte, like, not eq etc.
+    enum ComparisonOperator {
+        EQUALS = 0;
+    }
+
+    ComparisonOperator operator = 10; // field 10 in case we add more entities to query
+    // Next field number: 11
+}
+
+message ArtifactFilter {
+    oneof property {
+        string artifact_id = 1;
+    }
+}
+
+message TagFilter {
+    oneof property {
+        string tag_name = 1;
+    }
+}
+
+message PartitionFilter {
+    oneof property {
+        KeyValuePair key_val = 1;
+    }
+}
+
+message KeyValuePair {
+    string key = 1;
+    string value = 2;
+}
+
+message DatasetFilter {
+    oneof property {
+        string project = 1;
+        string name = 2;
+        string domain = 3;
+        string version = 4;
```

Comment:
Shouldn't the DatasetFilter be comprised of all these items and not be a `oneof`? Because you could ask for all datasets within the 'basemaps' project and in domain 'staging', right?

Reply:
Each of these `Filter`s are for a single property of that entity. See the message `SingleValueFilter` above which will contain one of these filters. The client will need to construct a `FilterExpression` that specifies a list of `SingleValueFilter` to combine. So in the case of the Datasets query you're asking about it would be something like:
```
FilterExpression {
    filters: [DatasetFilter.project='basemaps', Dataset.domain='...']
}
```

Reply:
The reason we want to represent this as `oneof` only instead of a message of many values is because later imagine you'd like to say something like `Dataset.version='1' OR `Dataset.version='2'`. The `SingleValueExpression` allows for that.

Reply:
Perhaps the name `DatasetPropertyFilter` would make more sense than just `DatasetFilter`? I'll make that change, let me know what you think.

Reply:
Ok, your responses make sense. And yes, `DatasetPropertyFilter` would be better I think.

Reply:
Ok, your responses make sense. And yes, DatasetPropertyFilter would be better I think.